File: Data Structures & Algorithms/daily-temperatures/submission-1.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int n = temperatures.size();
        vector<int> ans(n, 0);
        stack<pair<int, int>> stack; // {temp, idx}
        for (int i = 0; i < n; i++) {
            int cur = temperatures[i];
            while (!stack.empty() && cur > stack.top().first) {
                auto pair = stack.top();
                stack.pop();
                ans[pair.second] = i - pair.second;
            }
            stack.push({cur, i});
        }
        return ans;
    }
};

File: Data Structures & Algorithms/asteroid-collision/submission-1.py
class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        ans = []

        for val in asteroids: 
            while ans and val < 0 and ans[-1] > 0:
                diff = val + ans[-1]
                if diff < 0:
                    ans.pop()
                elif diff > 0:
                    val = 0
                else:
                    ans.pop()
                    val = 0
            if val:
                ans.append(val)
        return ans